### PR TITLE
sort and uniq imports and strings before returning them.

### PR DESCRIPTION
**Pull request by nevun**
attempt to fix issue #6461

### CODE REVIEW

File: binr/radiff2/radiff2.c [hunk 1]
```diff
@@ -364,52 +364,81 @@ static ut8 *slurp(RCore **c, const char *file, int *sz) {
...
+
 	r_list_foreach (list, iter, str) {
+		if (old && bs_cmp(old, str) == 0)
+			continue;
 		memcpy (ptr, str->string, str->length);
 		ptr += str->length;
 		*ptr++ = '\n';
+		old = str;
 	}
+	*ptr = 0;
```
Comment (nevun) [inline binr/radiff2/radiff2.c]: thatline is a pure bug

File: binr/radiff2/radiff2.c [hunk 2]
```diff
@@ -364,52 +364,81 @@ static ut8 *slurp(RCore **c, const char *file, int *sz) {
 	return (ut8*)r_file_slurp (file, sz);
 }
 
-// TODO: sort and uniq() https://github.com/radare/radare2/issues/6461
+static int import_cmp(const RBinImport* a, const RBinImport* b) {
+	return strcmp(a->name, b->name);
```
Comment (radare) [inline binr/radiff2/radiff2.c]: missing space before (

File: binr/radiff2/radiff2.c [hunk 3]
```diff
@@ -364,52 +364,81 @@ static ut8 *slurp(RCore **c, const char *file, int *sz) {
...
+	RBinImport *str, *old = NULL;
 	ut8 *buf, *ptr;
 
+	r_list_sort (list, (RListComparator)import_cmp);
+
 	*len = 0;
 
 	r_list_foreach (list, iter, str) {
-		*len += strlen (str->name) + 1;
+		if (!old || (old && import_cmp(old, str) != 0)) {
```
Comment (radare) [inline binr/radiff2/radiff2.c]: missing space before (

File: binr/radiff2/radiff2.c [hunk 4]
```diff
@@ -364,52 +364,81 @@ static ut8 *slurp(RCore **c, const char *file, int *sz) {
...
+			*len += strlen (str->name) + 1;
+			old = str;
+		}
 	}
 	ptr = buf = malloc (*len + 1);
```
Comment (radare) [inline binr/radiff2/radiff2.c]: check for null

File: binr/radiff2/radiff2.c [hunk 5]
```diff
@@ -364,52 +364,81 @@ static ut8 *slurp(RCore **c, const char *file, int *sz) {
...
 
+	old = NULL;
+
 	r_list_foreach (list, iter, str) {
+		if (old && import_cmp(old, str) == 0)
```
Comment (radare) [inline binr/radiff2/radiff2.c]: add braces

File: binr/radiff2/radiff2.c [hunk 6]
```diff
@@ -364,52 +364,81 @@ static ut8 *slurp(RCore **c, const char *file, int *sz) {
...
 	*ptr = 0;
 
 	*len = strlen ((const char *)buf);
 	return buf;
 }
 
-// TODO: sort and uniq() https://github.com/radare/radare2/issues/6461
+static int bs_cmp(const RBinString* a, const RBinString* b) {
+	int diff = a->length - b->length;
+	return diff == 0 ? strncmp(a->string, b->string, a->length) : diff;
```
Comment (radare) [inline binr/radiff2/radiff2.c]: missing space

File: binr/radiff2/radiff2.c [hunk 7]
```diff
@@ -419,7 +420,7 @@ static ut8 *get_strings(RCore *c, int *len) {
 	*len = 0;
 
 	r_list_foreach (list, iter, str) {
-		if (!old || (old && bs_cmp(old, str) != 0)) {
+		if (old == NULL || (old && bs_cmp (old, str) != 0)) {
```
Comment (radare) [inline binr/radiff2/radiff2.c]: Code style guide suggests the use of ! Instead of ==NULL

File: binr/radiff2/radiff2.c [hunk 8]
```diff
@@ -364,52 +364,83 @@ static ut8 *slurp(RCore **c, const char *file, int *sz) {
...
+	RBinImport *str, *old = NULL;
 	ut8 *buf, *ptr;
 
+	r_list_sort (list, (RListComparator)import_cmp);
+
 	*len = 0;
 
 	r_list_foreach (list, iter, str) {
-		*len += strlen (str->name) + 1;
+		if (old == NULL || (old && import_cmp (old, str) != 0)) {
```
Comment (radare) [inline binr/radiff2/radiff2.c]: Use !

File: binr/radiff2/radiff2.c [hunk 9]
```diff
@@ -364,52 +364,83 @@ static ut8 *slurp(RCore **c, const char *file, int *sz) {
...
+	RBinString *str, *old = NULL;
 	ut8 *buf, *ptr;
 
+	r_list_sort (list, (RListComparator)bs_cmp);
+
 	*len = 0;
 
 	r_list_foreach (list, iter, str) {
-		*len += str->length + 1;
+		if (old == NULL || (old && bs_cmp (old, str) != 0)) {
```
Comment (radare) [inline binr/radiff2/radiff2.c]: Use !

**Comment by radare** [comment]
Oh! damn, just noticed you are sorting by hand when there's already r_list_sort() which uses faster algorithms. Can you send another PR with that change? it will simplify the code and run faster. Also, having tests will be good for this

**Comment by nevun** [comment]
I did use r_list_sort().. What I am doing in the other loops is eliminating duplicates when producing the buffer that is to be returned.

**Comment by radare** [comment]
👍 go for the test :)

**Pull request merged**